build.rs

```diff
@@ -0,0 +1,9 @@
+fn main() {
+    println!("cargo:rerun-if-changed=build.rs");
+
+    // Pass in build-time environment variables, which could be used in
+    // crates by `env!` macros.
+    println!("cargo:rustc-env=PATH_HOSTNAMECTL=/usr/bin/hostnamectl");
+    println!("cargo:rustc-env=PATH_USERADD=/usr/sbin/useradd");
+    println!("cargo:rustc-env=PATH_PASSWD=/usr/bin/passwd");
```

Comment:
Actually, the approach already allows this for downstream packagers as far as I can tell. Distros only need to drop the absolute path, e.g. `export PATH_HOSTNAMECTL=hostnamectl` at build time, and relative paths / standard binary search `$PATH` will be used.

Reply:
Updated this part as @jeremycline suggested. Thanks.

Reply:
I think the non-absolute default variable is better, but what I meant (and didn't properly express) is I don't think we should have these configuration options at all. The operating systems already provide a well-known tunable for executable discovery (`PATH`) so providing a second way to do it feels unnecessary and potentially surprising.

If we add support for other tools some settings becomes irrelevant and confusing. I don't think FreeBSD has `hostnamectl`. From a library-user perspective, `set_hostname()` isn't abstracting how it sets the hostname as it takes `path_hostnamectl` as an argument, so it's not clear how the API can accommodate alternate tools. I think it would be better to drop these environment variables and document that tools need to be on the PATH.

The question remains on how you want to handle the library side of this. Does the library API let you select a backend to use, or do we make users select at compile time? I think it makes sense to design the API to allow callers to select a backend if we're providing a library.

Reply:
Thanks, @jeremyclient. I created a new [issue](https://github.com/Azure/azure-init/issues/89) for that.